### PR TITLE
Fix #340: Parse ASN1_TIME to struct tm

### LINKED ISSUE #340: Provide a public way to parse ASN1_GENERALIZEDTIME
The `ASN1_GENERALIZEDTIME` struct is used to return some important pieces of information related to OCSP, but as far as I can tell, there's no built-in way to convert it to a `time_t` or `tm`. This forces applications to parse the string themselves, which works, but is a pain. Since `asn1_generalizedtime_to_tm` exists, I don't see why it isn't public API.

One confusing point here: I'm not entirely clear on what time zone the returned `tm` is, but either ensuring it's UTC or providing a way to find out, and then documenting that either way, seems sane enough.

### PR DESCRIPTION
**Pull request by tmshort**
This works with ASN1_UTCTIME and ASN1_GENERALIZED_TIME

##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by tmshort** [comment]
Fixes #340

File: crypto/asn1/a_time.c [hunk 1]
```diff
@@ -130,7 +130,7 @@ int ASN1_TIME_set_string(ASN1_TIME *s, const char *str)
     return 1;
 }
 
-static int asn1_time_to_tm(struct tm *tm, const ASN1_TIME *t)
+int ASN1_TIME_to_tm(struct tm *tm, const ASN1_TIME *t)
```
Comment (mattcaswell) [inline crypto/asn1/a_time.c]: I'm wondering what order the params should be in. Recently we have tended to always put the primary object first with the function name also starting with the primary object name. But here the ASN1_TIME type comes second.

File: doc/man3/ASN1_TIME_set.pod [hunk 2]
```diff
@@ -13,6 +13,7 @@ ASN1_TIME_print, ASN1_TIME_diff - ASN.1 Time functions
  int ASN1_TIME_set_string(ASN1_TIME *s, const char *str);
  int ASN1_TIME_check(const ASN1_TIME *t);
  int ASN1_TIME_print(BIO *b, const ASN1_TIME *s);
+ int ASN1_TIME_to_tm(struct tm *tm, const ASN1_TIME *t);
```
Comment (mattcaswell) [inline doc/man3/ASN1_TIME_set.pod]: Needs to be added to the NAME section too.

File: doc/man3/ASN1_TIME_set.pod [hunk 3]
```diff
@@ -123,6 +129,9 @@ otherwise.
 ASN1_TIME_print() returns 1 if the time is successfully printed out and 0 if
 an error occurred (I/O error or invalid time format).
 
+ASN1_TIME_to_tm() returns 1 if the time is successfully parsed and 0 if an
+error occured (invalid time format).
+
 ASN1_TIME_diff() returns 1 for success and 0 for failure. It can fail if the
 pass ASN1_TIME structure has invalid syntax for example.
```
Comment (mattcaswell) [inline doc/man3/ASN1_TIME_set.pod]: Please can we add a HISTORY section and update the COPYRIGHT date.

File: crypto/asn1/a_time.c (hunk 1 above)
Comment (richsalz) [inline crypto/asn1/a_time.c]: i agree it should be flipped.

File: crypto/asn1/a_time.c [hunk 4]
```diff
@@ -152,9 +152,9 @@ int ASN1_TIME_diff(int *pday, int *psec,
                    const ASN1_TIME *from, const ASN1_TIME *to)
 {
     struct tm tm_from, tm_to;
-    if (!asn1_time_to_tm(&tm_from, from))
+    if (!ASN1_TIME_to_tm(&tm_from, from))
```
Comment (richsalz) [inline crypto/asn1/a_time.c]: since you're here, blank line before the declaration.

File: doc/man3/ASN1_TIME_set.pod (hunk 2 above)
Comment (richsalz) [inline doc/man3/ASN1_TIME_set.pod]: find-doc-nits is your friend. :)

File: doc/man3/ASN1_TIME_set.pod [hunk 5]
```diff
@@ -42,6 +43,11 @@ format. It will be of the format MMM DD HH:MM:SS YYYY [GMT], for example
 structure has invalid format it prints out "Bad time value" and returns
 an error.
 
+ASN1_TIME_to_tm() converts the time B<s> to the standard B<tm> structure.
```
Comment (richsalz) [inline doc/man3/ASN1_TIME_set.pod]: prose here says ```as``` but the prototype above says ```t```

File: crypto/asn1/a_time.c (hunk 1 above)
Comment (tmshort) [inline crypto/asn1/a_time.c]: Yup; I just un-static-fied the exiting function, but I agree, the arguments should be reversed.

Comment (tmshort) [inline crypto/asn1/a_time.c]: I will note that ASN1_TIME_print() is an anomaly (not suggesting this changes the need to reverse the arguments).

File: doc/man3/ASN1_TIME_set.pod (hunk 5 above)
Comment (tmshort) [inline doc/man3/ASN1_TIME_set.pod]: Changed `ASN1_TIME_to_tm()` everywhere to `s` to make it consistent with other ASN1_TIME functions.

**Comment by tmshort** [comment]
Fixed everything, I believe.

**Comment by tmshort** [comment]
rebase ping

File: doc/man3/ASN1_TIME_set.pod [hunk 6]
```diff
@@ -42,6 +43,11 @@ format. It will be of the format MMM DD HH:MM:SS YYYY [GMT], for example
...
+If B<s> is NULL, the the current time is converted. The output time is GMT.
```
Comment (mattcaswell) [inline doc/man3/ASN1_TIME_set.pod]: nit: then the

**Comment by tmshort** [comment]
Committed a doc tweak, which can be squashed.

**Comment by tmshort** [comment]
ping?

**Comment by mattcaswell** [comment]
@richsalz this needs your plus one

**Comment by tmshort** [comment]
> should we memset the 'struct tm' since there might be fields we don't know about? but approved.

The documentation explicitly states what fields are filled in, but yeah, I think it's a good idea.

**Comment by tmshort** [comment]
Fixed the merge conflict...

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**